### PR TITLE
Improvements 5

### PR DESCRIPTION
**Pull request by frankmayer**
### Summary of Changes
Type safe comparisons
Also a fix for a test with wrong expectations. (This should be looked at, carefully)
During the course of this fix, some PostGreSQL related code was also fixed.

### Testing Instructions
Code review only 
Caution. I also fixed a test that was having a wrong expectation and associated PostGreSQL code.

### CODE REVIEW

**Comment by frankmayer** [comment]
Pls don't merge yet. still some things to do...

**Comment by frankmayer** [comment]
@mbabker @andrepereiradasilva Ready for code review and merge
Caution. I changed a test expectation (and associated code) that looked just wrong to me. Please check.

File: Tests/QueryTest.php [hunk 1]
```diff
@@ -1749,7 +1749,7 @@ public function test__clone_array()
 		);
 
 		$this->assertFalse($baseElement === $cloneElement);
-		$this->assertTrue(count($cloneElement->testArray) == 0);
+		$this->assertTrue(count($cloneElement->testArray) === 0);
```
Comment (andrepereiradasilva) [inline Tests/QueryTest.php]: small thing. isn't it faster `$cloneElement->testArray === array()`

File: src/Pgsql/PgsqlDriver.php [hunk 2]
```diff
@@ -509,7 +510,7 @@ public function sqlValue($columns, $field_name, $field_value)
 			case 'smallint':
 			case 'serial':
 			case 'numeric,':
-				$val = strlen($field_value) == 0 ? 'NULL' : $field_value;
+				$val = strlen($field_value) === 0 ? 'NULL' : $field_value;
```
Comment (andrepereiradasilva) [inline src/Pgsql/PgsqlDriver.php]: `$field_value === ''`?

File: src/Postgresql/PostgresqlDriver.php [hunk 3]
```diff
@@ -1016,7 +1016,7 @@ public function sqlValue($columns, $field_name, $field_value)
 			case 'smallint':
 			case 'serial':
 			case 'numeric,':
-				$val = strlen($field_value) == 0 ? 'NULL' : $field_value;
+				$val = strlen($field_value) === 0 ? 'NULL' : $field_value;
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlDriver.php]: `$field_value === ''`?

File: src/Postgresql/PostgresqlDriver.php [hunk 4]
```diff
@@ -1557,7 +1557,7 @@ public function updateObject($table, &$object, $key, $nulls = false)
 			}
 
 			// Only process scalars that are not internal fields.
-			if (is_array($v) or is_object($v) or $k[0] == '_')
+			if (is_array($v) or is_object($v) or $k[0] === '_')
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlDriver.php]: `||`?

File: src/Postgresql/PostgresqlImporter.php [hunk 5]
```diff
@@ -348,9 +348,9 @@ protected function getAlterColumnSql($table, \SimpleXMLElement $field)
 
 		$sql = ' TYPE ' . $fType;
 
-		if ($fNull == 'NO')
+		if ($fNull === 'NO')
 		{
-			if (in_array($fType, $blobs) || $fDefault === null)
+			if (in_array($fType, $blobs, true) || $fDefault == null)
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlImporter.php]: why `$fDefault == null`?

File: Tests/QueryTest.php (hunk 1 above)
Comment (frankmayer) [inline Tests/QueryTest.php]: Don't know. Need to check.

File: src/Pgsql/PgsqlDriver.php (hunk 2 above)
Comment (frankmayer) [inline src/Pgsql/PgsqlDriver.php]: Ah, yes. had those done, but reverted some changes after having the issues with the problematic test.

File: src/Postgresql/PostgresqlImporter.php (hunk 5 above)
Comment (frankmayer) [inline src/Postgresql/PostgresqlImporter.php]: Same as above.. will fix. Thanks

File: src/DatabaseDriver.php [hunk 6]
```diff
@@ -1838,7 +1838,7 @@ public function updateObject($table, &$object, $key, $nulls = false)
 			}
 
 			// Set the primary key to the WHERE clause instead of a field to update.
-			if (in_array($k, $key))
+			if (in_array($k, $key, true))
```
Comment (andrepereiradasilva) [inline src/DatabaseDriver.php]: just one thing to discuss.

when possible strict checking a php function result?
seems to be a little faster.
i guess it makes sense since it does not have to check for a lot of values: `1`, `-1`, `'1'`, `'-1'`, `'php'` (no idea why :smile:), `true` (see http://php.net/manual/en/types.comparisons.php), only `true`.

at least with this code seems faster
```php
$array = ['a', 'b'];

// Loose checking in_array result
$start = microtime(true);
for ($i = 1; $i <= 1000000; $i++)
{
	if (in_array('a', $array, true))
	{ /* do nothing; */ }
}
echo '[' . number_format(((microtime(true) - $start) * 1000), 3) . ' ms]' .PHP_EOL;

// Strict checking in_array result
$start = microtime(true);
for ($i = 1; $i <= 1000000; $i++)
{
	if (in_array('a', $array, true) === true)
	{ /* do nothing; */ }
}
echo '[' . number_format(((microtime(true) - $start) * 1000), 3) . ' ms]' .PHP_EOL;
```
Result:
- Loose: 195.858 ms
- Strict: 133.810 ms

Not sure about this, but jsut one thing i noticed.

File: src/Mysqli/MysqliImporter.php [hunk 7]
```diff
@@ -287,9 +287,9 @@ protected function getColumnSql(\SimpleXMLElement $field)
 
 		$sql = $this->db->quoteName($fName) . ' ' . $fType;
 
-		if ($fNull == 'NO')
+		if ($fNull === 'NO')
 		{
-			if (in_array($fType, $blobs) || $fDefault === null)
+			if (in_array($fType, $blobs, true) || $fDefault === null)
```
Comment (andrepereiradasilva) [inline src/Mysqli/MysqliImporter.php]: `if ($fDefault === null || in_array($fType, $blobs, true))`?

File: src/Postgresql/PostgresqlDriver.php [hunk 8]
```diff
@@ -528,7 +528,7 @@ public function getTableKeys($table)
 		// To check if table exists and prevent SQL injection
 		$tableList = $this->getTableList();
 
-		if ( in_array($table, $tableList) )
+		if ( in_array($table, $tableList, true) )
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlDriver.php]: remove spaces

File: src/Postgresql/PostgresqlDriver.php [hunk 9]
```diff
@@ -593,7 +593,7 @@ public function getTableSequences($table)
...
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlDriver.php]: remove spaces

File: src/Postgresql/PostgresqlDriver.php [hunk 10]
```diff
@@ -852,7 +852,7 @@ public function renameTable($oldTable, $newTable, $backup = null, $prefix = null
 		$tableList = $this->getTableList();
 
 		// Origin Table does not exist
-		if ( !in_array($oldTable, $tableList) )
+		if ( !in_array($oldTable, $tableList, true) )
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlDriver.php]: remove spaces

File: src/Postgresql/PostgresqlQuery.php [hunk 11]
```diff
@@ -715,7 +715,7 @@ public function processLimit($query, $limit, $offset = 0)
 	 */
 	public function dateAdd($date, $interval, $datePart)
 	{
-		if (substr($interval, 0, 1) != '-')
+		if (substr($interval, 0, 1) !== '-')
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlQuery.php]: `strpos($interval, '-') !== 0` or maybe even `$interval[0] !== '-'` ?

File: src/Postgresql/PostgresqlImporter.php [hunk 12]
```diff
@@ -338,7 +338,7 @@ protected function getAlterColumnSql($table, \SimpleXMLElement $field)
 	{
 		// TODO Incorporate into parent class and use $this.
 		$blobs = array('text', 'smalltext', 'mediumtext', 'largetext');
-
+		var_dump($field);
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlImporter.php]: ?

File: src/Postgresql/PostgresqlImporter.php [hunk 13]
```diff
@@ -392,6 +392,7 @@ protected function getColumnSql(\SimpleXMLElement $field)
 	{
 		// TODO Incorporate into parent class and use $this.
 		$blobs = array('text', 'smalltext', 'mediumtext', 'largetext');
+		var_dump($field);
```
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlImporter.php]: ?

File: src/Sqlsrv/SqlsrvDriver.php [hunk 14]
```diff
@@ -508,6 +508,9 @@ public function insertObject($table, &$object, $key = null)
 		$values = array();
 		$tableColumns = $this->getTableColumns($table);
 		$statement = 'INSERT INTO ' . $this->quoteName($table) . ' (%s) VALUES (%s)';
+		var_dump($table);
+		var_dump($object);
+		var_dump($key);
```
Comment (andrepereiradasilva) [inline src/Sqlsrv/SqlsrvDriver.php]: ?

File: src/DatabaseDriver.php (hunk 6 above)
Comment (frankmayer) [inline src/DatabaseDriver.php]: Yes, have thought about this too in the past, but didn't know that it had such a performance impact. Definitely yes.

Comment (frankmayer) [inline src/DatabaseDriver.php]: Hom, taking your code and reversing the order of the tests leads to non-strict checking being faster most of the times...
I'd say non-conclusive on the performance gain. On generally doing this for stricter comparisons is a thing to discuss, though.

File: src/Postgresql/PostgresqlQuery.php (hunk 11 above)
Comment (frankmayer) [inline src/Postgresql/PostgresqlQuery.php]: Yes, that would be part of another PR

File: src/Postgresql/PostgresqlImporter.php (hunk 12 above)
Comment (frankmayer) [inline src/Postgresql/PostgresqlImporter.php]: Trying to figure something out. pls ignore. Will let you know

File: src/DatabaseDriver.php (hunk 6 above)
Comment (andrepereiradasilva) [inline src/DatabaseDriver.php]: i think the performance varies a lot depending on the result of the functions.
more tests are needed to be sure of that
anyway from a code correctness perspective it would be also nice i guess

File: src/Mysqli/MysqliImporter.php (hunk 7 above)
Comment (frankmayer) [inline src/Mysqli/MysqliImporter.php]: the check against null costs less

Comment (andrepereiradasilva) [inline src/Mysqli/MysqliImporter.php]: i know :) that's why i'm proposing it should be first.

File: src/Postgresql/PostgresqlQuery.php (hunk 11 above)
Comment (andrepereiradasilva) [inline src/Postgresql/PostgresqlQuery.php]: ok

File: src/DatabaseDriver.php (hunk 6 above)
Comment (andrepereiradasilva) [inline src/DatabaseDriver.php]: but i guess that would be other PR's

File: Tests/QueryTest.php (hunk 1 above)
Comment (andrepereiradasilva) [inline Tests/QueryTest.php]: check this

```php
$arrays = [
	'array1' => [],
	'array2' => ['a', 'b'],
	'array3' => ['a' => 'a', 'b' => 'b'],
];

foreach ($arrays as $key => $array)
{
	$start = microtime(true);
	for ($i = 1; $i <= 1000000; $i++)
	{
		count($array) === 0 ? '' : '';
	}
	echo'[' . number_format(((microtime(true) - $start) * 1000), 3) . ' ms] ' . $key . ' count' .PHP_EOL;

	$start = microtime(true);
	for ($i = 1; $i <= 1000000; $i++)
	{
		$array === [] ? '' : '';
	}
	echo '[' . number_format(((microtime(true) - $start) * 1000), 3) . ' ms] ' . $key . ' empty compare' .PHP_EOL;
}
```

Result:
[166.196 ms] array1 count
[127.573 ms] array1 empty compare

[167.252 ms] array2 count
[95.666 ms] array2 empty compare

[68.628 ms] array3 count
[53.312 ms] array3 empty compare

File: src/Mysqli/MysqliImporter.php (hunk 7 above)
Comment (frankmayer) [inline src/Mysqli/MysqliImporter.php]: Oh :smile_cat:  misread the comment... I had done the same change at another statement

File: src/DatabaseDriver.php (hunk 6 above)
Comment (frankmayer) [inline src/DatabaseDriver.php]: @mbabker any thoughts on that?

Comment (frankmayer) [inline src/DatabaseDriver.php]: other PRs definitely... :)

**Comment by mbabker** [comment]
Someone just tell me when this PR is good to review/merge 😉

**Comment by frankmayer** [comment]
:) @mbabker there were some issues I had to check. If @andrepereiradasilva thinks it's fine then I would say it's ok for you to check and merge.

**Comment by andrepereiradasilva** [comment]
sure the rest can be in another PRs

**Comment by frankmayer** [comment]
@mbabker this is ready to be merged, then.
